vars/cloudFoundryDeploy.groovy

```diff
@@ -0,0 +1,166 @@
+            } else {
+                utils.unstash 'pipelineConfigAndTests'
+            }
+            smokeTest = '--smoke-test $(pwd)/' + config.smokeTestScript
+            sh "chmod +x ${config.smokeTestScript}"
+
+            echo "[${STEP_NAME}] CF native deployment (${config.deployType}) with cfAppName=${config.cloudFoundry.appName}, cfManifest=${config.cloudFoundry.manifest}, smokeTestScript=${config.smokeTestScript}"
+
+            dockerExecute (
+                dockerImage: config.dockerImage,
+                dockerWorkspace: config.dockerWorkspace,
+                stashContent: config.stashContent,
+                dockerEnvVars: [CF_HOME:"${config.dockerWorkspace}", CF_PLUGIN_HOME:"${config.dockerWorkspace}", STATUS_CODE: "${config.smokeTestStatusCode}"]
+            ) {
+                deployCfNative(config)
+            }
+
+            return
+        }
+    }
+}
+
+def findMtar(){
+    def mtarPath = ''
+    def mtarFiles = findFiles(glob: '**/target/*.mtar')
+
+    if(mtarFiles.length > 1){
+        error 'Found multiple *.mtar files, please specify file via mtaPath parameter! ${mtarFiles}'
+    }
+    if(mtarFiles.length == 1){
+        return mtarFiles[0].path
+    }
+    error 'No *.mtar file found!'
+}
+
+def deployCfNative (config) {
+    withCredentials([usernamePassword(
+        credentialsId: config.cloudFoundry.credentialsId,
+        passwordVariable: 'password',
+        usernameVariable: 'username'
+    )]) {
+        def deployCommand = 'push'
+        if (config.deployType == 'blue-green') {
+            deployCommand = 'blue-green-deploy'
+        } else {
+            config.smokeTest = ''
+        }
+
+        // check if appName is available
+        if (config.cloudFoundry.appName == null || config.cloudFoundry.appName == '') {
+            if (fileExists(config.cloudFoundry.manifest)) {
+                def manifest = readYaml file: config.cloudFoundry.manifest
+                if (!manifest || !manifest.applications || !manifest.applications[0].name)
+                    error "[${STEP_NAME}] ERROR: No appName available in manifest ${config.cloudFoundry.manifest}."
+
+            } else {
+                error "[${STEP_NAME}] ERROR: No manifest file ${config.cloudFoundry.manifest} found."
+            }
+        }
+
+        sh """#!/bin/bash
+            set +x
+            export HOME=${config.dockerWorkspace}
+            cf login -u \"${username}\" -p '${password}' -a ${config.cloudFoundry.apiEndpoint} -o \"${config.cloudFoundry.org}\" -s \"${config.cloudFoundry.space}\"
+            cf plugins
+            cf ${deployCommand} ${config.cloudFoundry.appName?"\"${config.cloudFoundry.appName}\"":''} -f \"${config.cloudFoundry.manifest}\" ${config.smokeTest}"""
+        def retVal = sh script: "cf app \"${config.cloudFoundry.appName}-old\"", returnStatus: true
+        if (retVal == 0) {
+            sh "cf delete \"${config.cloudFoundry.appName}-old\" -r -f"
+        }
+        sh "cf logout"
```

Comment:
what is the reason for having cf logout in an other shell instead of having it the previous shell? In case the reason is: 'we would like to ensure that the logout happens in any case' we could also use good old trap

Reply:
helps that it is called also for retVal !=0.